lib/Runtime/Base/ScriptContext.cpp

```diff
@@ -2178,28 +2179,71 @@ namespace Js
+        ULONG compressedBufferByteCount = blockByteCount - sizeof(uint);
 
         // The contract for this bytecode buffer is that it is available as long as we have this ScriptContext.
         // We will use this buffer as the string table needed to back the deferred stubs as well as bytecode
         // for defer deserialized functions.
         // TODO: This, better.
         ArenaAllocator* alloc = this->SourceCodeAllocator();
-        byte* buffer = AnewArray(alloc, byte, byteCount);
+        size_t decompressedBufferByteCount = 0;
+        byte* decompressedBuffer = nullptr;
 
-        if (buffer == nullptr)
+        if (CONFIG_FLAG(CompressParserStateCache))
         {
-            return E_FAIL;
-        }
+            BEGIN_TEMP_ALLOCATOR(tempAllocator, this, _u("ByteCodeSerializer"));
+            {
+                byte* compressedBuffer = AnewNoThrowArray(tempAllocator, byte, compressedBufferByteCount);
+                if (compressedBuffer == nullptr)
+                {
+                    hr = E_FAIL;
+                    goto ExitTempAllocator;
+                }
 
-        hr = pDataCache->ReadArray(buffer, byteCount);
-        if (FAILED(hr))
+                hr = pDataCache->ReadArray(compressedBuffer, compressedBufferByteCount);
+                if (FAILED(hr))
+                {
+                    OUTPUT_TRACE_DEBUGONLY(Js::DataCachePhase, _u(" Failed to read compressed parser state cache (wanted %lu bytes) (hr = 0x%08lx) for '%s'\n"), compressedBufferByteCount, hr, url);
+                    goto ExitTempAllocator;
+                }
+
+                hr = Js::CompressionUtilities::DecompressBuffer(alloc, compressedBuffer, compressedBufferByteCount, &decompressedBuffer, &decompressedBufferByteCount);
+                if (FAILED(hr))
+                {
+                    OUTPUT_TRACE_DEBUGONLY(Js::DataCachePhase, _u(" Failed to decompress parser state cache (hr = 0x%08lx) for '%s'\n"), hr, url);
+                    goto ExitTempAllocator;
+                }
+
+                OUTPUT_TRACE_DEBUGONLY(Js::DataCachePhase, _u(" Decompressed parser state cache %lu -> %lu bytes (%.2f%%) to stream for '%s'\n"), compressedBufferByteCount, decompressedBufferByteCount, (double)compressedBufferByteCount / decompressedBufferByteCount * 100.0, url);
+            }
+ExitTempAllocator:
+            END_TEMP_ALLOCATOR(tempAllocator, this);
+
+            if (FAILED(hr))
+            {
+                AdeleteArray(alloc, decompressedBufferByteCount, decompressedBuffer);
+                return hr;
+            }
+        }
+        else
         {
-            OUTPUT_TRACE_DEBUGONLY(Js::DataCachePhase, _u(" Failed to read parser state cache (wanted %lu bytes, got %lu bytes) (hr = 0x%08lx) for '%s'\n"), byteCount, pDataCache->BytesWrittenInBlock(), hr, url);
-            return hr;
+            // We didn't compress the parser state cache so don't decompress it, just read the buffer.
```

Comment:
Will this lead to issues if you run chakra in multiple configurations on the one machine? e.g. you run it in compressed mode, then toggle jscript.config and run in non-compressed mode and go to the same site, will that attempt to treat compressed data as raw data and get confused? Is there any safety net there?

Reply:
Yes, doing that we will confuse the type of data and might try to interpret bytecode as compressed data or vice versa. We should gracefully fail in either case. The compression API will fail to decompress bytecode with a failing HRESULT and the bytecodeserializer will fail to deserialize compressed data, as well. We shouldn't crash but we will end up discarding the saved parser state cache in both cases, overwriting with a new one.

In retail scenarios this is a non-issue as there is a version check in the SimpleDataCacheWrapper which will throw out data in the cache from a previous version of Chakra.

I considered adding a couple of bytes of metadata to the cache to indicate what settings the parser state cache was created with so we wouldn't even attempt to get confused about the data. Wasn't clear what we would want to do here, though, this should only be a debugging scenario. If we try to read a parser state cache which was compressed but we are now running with compression disabled should we decompress the cache anyway? Return a failure?

Reply:
I think discarding out-of-date results is fine, just wanted to make sure it did in fact behave that way.